runtime/blake2.c

```diff
@@ -0,0 +1,245 @@
+    c += d;                                                                 \
+    b = ROTR64(b ^ c, 63);                                                  \
+  } while(0)                                                                \
+
+static void
+caml_BLAKE2Compress(struct BLAKE2_context * s,
+                    const unsigned char * data, unsigned int numbytes,
+                    int is_last_block)
+{
+  uint64_t v0, v1, v2, v3, v4, v5, v6, v7,
+           v8, v9, v10, v11, v12, v13, v14, v15;
+  uint64_t m[16];
+
+  /* Update the length */
+  s->len[0] += numbytes;
+  if (s->len[0] < numbytes) s->len[1]++; /* carry */
+  /* Initialize work space */
+  v0 = s->h[0];  v1 = s->h[1];
+  v2 = s->h[2];  v3 = s->h[3];
+  v4 = s->h[4];  v5 = s->h[5];
+  v6 = s->h[6];  v7 = s->h[7];
+  v8 = caml_BLAKE2_iv[0];  v9 = caml_BLAKE2_iv[1];
+  v10 = caml_BLAKE2_iv[2]; v11 = caml_BLAKE2_iv[3];
+  v12 = caml_BLAKE2_iv[4] ^ s->len[0];  v13 = caml_BLAKE2_iv[5] ^ s->len[1];
+  v14 = is_last_block ? ~ caml_BLAKE2_iv[6] : caml_BLAKE2_iv[6];
+  v15 = caml_BLAKE2_iv[7];
+  /* Convert data to 16 64-bit words */
+  for (int i = 0; i < 16; i++) {
+    m[i] = U8TO64LE(data + i * 8);
+  }
+  /* Twelve rounds of mixing */
+  for (int i = 0; i < 12; i++) {
+    const uint8_t * sigma = BLAKE2_sigma[i];
+    MIX2B(v0, v4, v8,  v12, m[sigma[0]],  m[sigma[1]]);
+    MIX2B(v1, v5, v9,  v13, m[sigma[2]],  m[sigma[3]]);
+    MIX2B(v2, v6, v10, v14, m[sigma[4]],  m[sigma[5]]);
+    MIX2B(v3, v7, v11, v15, m[sigma[6]],  m[sigma[7]]);
+    MIX2B(v0, v5, v10, v15, m[sigma[8]],  m[sigma[9]]);
+    MIX2B(v1, v6, v11, v12, m[sigma[10]], m[sigma[11]]);
+    MIX2B(v2, v7, v8,  v13, m[sigma[12]], m[sigma[13]]);
+    MIX2B(v3, v4, v9,  v14, m[sigma[14]], m[sigma[15]]);
+  }
+  /* Update state  */
+  s->h[0] ^= v0 ^ v8;   s->h[1] ^= v1 ^ v9;
+  s->h[2] ^= v2 ^ v10;  s->h[3] ^= v3 ^ v11;
+  s->h[4] ^= v4 ^ v12;  s->h[5] ^= v5 ^ v13;
+  s->h[6] ^= v6 ^ v14;  s->h[7] ^= v7 ^ v15;
+}
+
+CAMLexport void
+caml_BLAKE2Init(struct BLAKE2_context * s,
+                int hashlen,
+                int keylen, const unsigned char * key)
+{
+  CAMLassert (0 < hashlen && hashlen <= 64);
+  CAMLassert (0 <= keylen);
+  for (int i = 0; i < 8; i++) s->h[i] = caml_BLAKE2_iv[i];
+  s->h[0] ^= 0x01010000 | (keylen << 8) | hashlen;
+  s->len[0] = s->len[1] = 0;
+  s->numbytes = 0;
+  /* If key was supplied, pad to 128 bytes and prepend to message */
+  if (keylen > 0) {
+    if (keylen > 64) keylen = 64;
+    memcpy(s->buffer, key, keylen);
+    memset(s->buffer + keylen, 0, BLAKE2_BLOCKSIZE - keylen);
+    s->numbytes = BLAKE2_BLOCKSIZE;
+  }
+}
+
+CAMLexport void
+caml_BLAKE2Update(struct BLAKE2_context * s,
+                  const unsigned char * data, uintnat len)
+{
+  /* If data was left in buffer, pad it with fresh data and compress */
+  if (s->numbytes > 0) {
+    int n = BLAKE2_BLOCKSIZE - s->numbytes;
+    if (len <= n) {
```

Comment:
`n` is implicitly converted to an unsigned int here, I’m wondering whether we want to make `n` unsigned, along with the `numbytes` field of the state. I do not have a strong opinion about this, since it does not seem to add many guarantees to the code.

Reply:
Now unsigned (`size_t`).